torch/csrc/distributed/c10d/socket.cpp

```diff
@@ -0,0 +1,818 @@
+}
+
+inline void setSocketError(int val) noexcept {
+  ::WSASetLastError(val);
+}
+
+#else
+
+const auto pollFd = ::poll;
+
+const auto getSocketOption = ::getsockopt;
+const auto setSocketOption = ::setsockopt;
+
+inline std::error_code getSocketError() noexcept {
+  return lastError();
+}
+
+inline void setSocketError(int val) noexcept {
+  errno = val;
+}
+
+#endif
+
+// Suspends the current thread for one second.
+void delay() {
+  constexpr std::chrono::seconds pause{1};
+
+#ifdef _WIN32
+  std::this_thread::sleep_for(pause);
+#else
+  ::timespec req{};
+  req.tv_sec = pause.count();
+
+  // The C++ Standard does not specify whether `thread::sleep_for()` should be
+  // signal-aware; therefore, we use the `nanosleep()` syscall.
+  if (::nanosleep(&req, nullptr) != 0) {
+    std::error_code err = getSocketError();
+    // We don't care about error conditions other than EINTR since a failure
+    // here is not critical.
+    if (err == std::errc::interrupted) {
+      throw std::system_error{err};
+    }
+  }
+#endif
+}
+
+class SocketListenOp;
+class SocketConnectOp;
+} // namespace
+
+class SocketImpl {
+  friend class SocketListenOp;
+  friend class SocketConnectOp;
+
+ public:
+#ifdef _WIN32
+  using Handle = SOCKET;
+#else
+  using Handle = int;
+#endif
+
+#ifdef _WIN32
+  static constexpr Handle invalid_socket = INVALID_SOCKET;
+#else
+  static constexpr Handle invalid_socket = -1;
+#endif
+
+  explicit SocketImpl(Handle hnd) noexcept
+      : hnd_{hnd} {}
+
+  SocketImpl(const SocketImpl& other) = delete;
+
+  SocketImpl& operator=(const SocketImpl& other) = delete;
+
+  SocketImpl(SocketImpl&& other) noexcept = delete;
+
+  SocketImpl& operator=(SocketImpl&& other) noexcept = delete;
+
+  ~SocketImpl();
+
+  std::unique_ptr<SocketImpl> accept() const;
+
+  void closeOnExec() noexcept;
+
+  void enableNonBlocking();
+
+  void disableNonBlocking();
+
+  bool enableNoDelay() noexcept;
+
+  bool enableDualStack() noexcept;
+
+#ifdef _WIN32
+  bool enableExclusiveAddressUse() noexcept;
+#else
+  bool enableAddressReuse() noexcept;
+#endif
+
+  std::uint16_t getPort() const;
+
+  Handle handle() const noexcept {
+    return hnd_;
+  }
+
+ private:
+  bool setSocketFlag(int level, int optname, bool value) noexcept;
+
+  Handle hnd_;
+};
+
+SocketImpl::~SocketImpl() {
+#ifdef _WIN32
+  ::closesocket(hnd_);
+#else
+  ::close(hnd_);
+#endif
+}
+
+std::unique_ptr<SocketImpl> SocketImpl::accept() const {
+  ::sockaddr_storage addr_s{};
+
+  auto addr_ptr = reinterpret_cast<::sockaddr*>(&addr_s);
+
+  ::socklen_t addr_len = sizeof(addr_s);
+
+  Handle hnd = ::accept(hnd_, addr_ptr, &addr_len);
+  if (hnd == invalid_socket) {
+    std::error_code err = getSocketError();
+    if (err == std::errc::interrupted) {
+      throw std::system_error{err};
+    }
+
+    std::string msg{};
+    if (err == std::errc::invalid_argument) {
+      msg = fmt::format("The server socket on {} is not listening for connections.", *this);
```

Comment:
curious, how did we decide to use this error message for `invalid_argument` error instead of saying "invalid argument"? Is this the convention for socket error handling?

Reply:
Below an excerpt from `man 2 accept`:

```
RETURN VALUE
  ...
  EINVAL Socket is not listening for connections, or addrlen is invalid (e.g., is negative).
```

Since we know for sure that `addrlen` is valid (see line 183), our error message correctly conveys the information.